File: trivial-compression/trivial_compression.py
class CompressedGene:
    
    def __init__(self, gene:str) -> None: 
        self._compress(gene)

    def _compress(self, gene: str) -> None:
        # _ indicates a 'private' method
        # Tip: using double leading underscores within a class will make Python 'name mangle' 
        # the implementation name. Meaning it'd add a salt to avoid it from being easily discoverable 
        # by other classes.

        self.bit_string: int = 1

        for nucleotide in gene.upper():
            self.bit_string <<=2 # shif left two bits
            if nucleotide == "A": # change last two bits to 00
                self.bit_string |= 0b00
            elif nucleotide == "C": # change last two bits to 00
                self.bit_string |= 0b01
            elif nucleotide == "G": # change last two bits to 00
                self.bit_string |= 0b10
            elif nucleotide == "T": # change last two bits to 00
                self.bit_string |= 0b11
            else:
                raise ValueError("Invalid Nucleotid: {}".format(nucleotide))
    
    def _decompress(self) -> str:
        gene: str = ""

        for i in range(0, self.bit_string.bit_length() -1, 2): #-1 to exclude sentinel
            bits: int = self.bit_string >> i & 0b11 # get just 2 relevant bits
            if bits == 0b00:
                gene += "A"
            elif bits == 0b01:
                gene += "C"
            elif bits == 0b10:
                gene += "G"
            elif bits == 0b11:
                gene += "T"
            else:
                raise ValueError("Invalid bits: {}".format(bits))

        return gene[::-11] # reverses string by slicing backwards
    
    def __str__(self) -> str:
        return self._decompress()

if __name__ == "__main__":
    from sys import getsizeof
    original: str = "TAGGGGAAATTAAAATGGGGAAATTTTAGAGGGAGAGGATTT" * 100
    print("original is {} bytes".format(getsizeof(original)))
    compressed: CompressedGene = CompressedGene(original)
    print("compressed is {} bytes".format(getsizeof(compressed)))
    print(compressed)
    decompressed: CompressedGene = compressed._decompress()
    print(decompressed)
    print("original and decompressed are the same: {}".format(original == decompressed))

File: trivial-compression/trivial_compression.js
function range(start, stop, step) {
    if (typeof stop == 'undefined') {
        // one param defined
        stop = start;
        start = 0;
    }

    if (typeof step == 'undefined') {
        step = 1;
    }

    if ((step > 0 && start >= stop) || (step < 0 && start <= stop)) {
        return [];
    }

    var result = [];
    for (var i = start; step > 0 ? i < stop : i > stop; i += step) {
        result.push(i);
    }

    return result;
};

class CompressedGene {
    constructor(gene) {
        this._compress(gene);
        
    }

    _compress(gene) {

        this.bit_string = 1;

        gene = gene.toUpperCase();

        for (let nucleotide of gene) {
            this.bit_string <<=2
            if (nucleotide === "A") {
                this.bit_string |= 0b00
            } else if (nucleotide === "C") {
                this.bit_string |= 0b01
            } else if (nucleotide === "G") {
                this.bit_string |= 0b10
            } else if (nucleotide === "T") {
                this.bit_string |= 0b11
            } else {
                console.log("Invalid Nucleotoid")
            }
        }
        console.log("Final Bit String ", this.bit_string.length)
    }

    _decompress() {
        //encoded(encodedGene >>> 0).toString(2)
        let gene = "";
        console.log(this.bit_string.length)
        let rango = range(0, this.bit_string.toString().length)
        
        for (let i of rango) {
            
            let bits = this.bit_string >> i & 0b00

            if (bits == Number("0b00")) {
                gene += "A"
            } else if (bits ===Number("0b01")) {
                gene += "C"
            } else if (bits == Number("0b10")) {
                gene += "G"
            } else if (bits == Number("0b11")) {
                gene += "T"
            } else {
                console.log("Invalid bits")
            }
        }

        return gene
    }
}

let original = "TAGGGGAAATTAAAATGGGGAAATTTTAGAGGGAGAGGATTT".repeat(100);
let compressed = new CompressedGene(original)
console.log(compressed)
let decompressed = compressed._decompress();
console.log(decompressed)